### ISSUE TITLE
Manage encoding files by extension of filenames

### ISSUE THREAD

**Issue by FrancisFaure** [Microsoft/vscode#4891]
Hello "VS Code" team,
- I do a language extension for "VS Code", 
  (https://github.com/FrancisFaure/vfp_tmlanguage_generator)
  (Visual FoxPro)
- I search a setting for the "codepage" on file extensions files (.PRG) on "Windows 1252" and not "UTF-8" by default
- I try (without succes to add "files.encoding": "windows1252" in "package.json"
- "VS Code" website say  : 
  users can change all codepage "globally" or codepage on "workspace"
  (https://code.visualstudio.com/docs/editor/codebasics#_file-encoding-support) 

But in the same directory we can have files in utf8 and others in "Western" (Windows 1252) codepage

I tried to add "files.encoding": "windows1252" in my language pack (package.json) without success..

I googled without solutions

sorry for my bad english

With best regards 
Francis

**Comment by bpasero** [comment]
Currently you can only configure encoding per workspace folder.

**Comment by FrancisFaure** [comment]
Thank for the quick answer... 
Have you plan to manage encoding files by extension of filenames ?
with by example add property "files.encoding" in file "package.json" of VS Code extension language.
Because same folder could have files encoded 1252 and others utf8
with best regards
Francis

**Comment by FrancisFaure** [comment]
please, 
have you planned an VSCode update for allow add "files.encoding": "windows1252" in extension language package.json file?
or a way for open files by extension of filenames encoded 1252 and not utf8 (we need to open *.PRG files in 1252)
Thanks

**Comment by buzzzzer** [comment]
Any news for files.encoding in extention API or by fileext ?
On my project 3 different encodings in over 1000 files. 
And reopen the file each time - just killing me...
Please schedule this feature..;(

**Comment by cheery** [comment]
I am also affected by this issue.

Please fix it.

**Comment by bpasero** [comment]
https://github.com/Microsoft/vscode/issues/1587

**Comment by fc89** [comment]
@bpasero That didn't fix the issue. We can now change editor.* settings but not the encoding setting.

**Comment by ianmaurinsoucy** [comment]
I am also affected by this issue.  I have many files in the same folders with different encoding...  I can't use VSCode because of this.

**Comment by FrancisFaure** [comment]
I left my volunteer work on VS Code because of this encoding problem that was ignored. cordially

**Comment by Somnium7** [comment]
We also need possibility to set different encoding for different file extensions. Maybe it's possible to write an extension for this?

**Comment by FrancisFaure** [comment]
to "Somnium7", 
Hi
I work a lot (volunteer) on VS Code for manage encoding files by extension of filenames (in example like .HTML -> UTF8 and .PRG -> 1512 in the same directory)
But Microsoft ignore this request : it seem he want only manage UTF8 encoding or encoding per directories,
(Or bpassero haven't understood my request with my bad english)
The solution can be only an new line in package.json file or other way. but ignored. 
if you find a solution for "set different encoding for different file extensions" : I'll happy
with best regards
Francis

**Comment by freMea** [comment]
I tried to add 

    "[bat]": {
        "files.encoding": "cp437",
        "files.autoGuessEncoding": false,
        "editor.fontFamily": "Consolas, 'Courier New', monospace"
    }

to the settings.json

`files.encoding` and `files.autoGuessEncoding` triggers an issue displayed in the tooltip on mouseover: _Unknown identifier. Use language identifiers._

If it is impossible now to set those settings as per language rule, please fix it.

_VS Code 1.11.2 on Win 10 x64_

**Comment by z639** [comment]
What I find funny is that my old>old editor of choice (Notepad++) is able to auto detect the file encoding and yet VS Code with all its bells and whistles that make it a far better editor can't even manage to do that same. 

At the very least you would think that we would be able to tell the editor to use X encoding for Y file type.

**Comment by jhonrocha** [comment]
This problem is still not solved? It is the biggest miss of Code right now ... Please, take it seriously for a moment. I never lost an afternoon replacing bad chars due a miss while saving a file on Sublime ...

**Comment by freMea** [comment]
Guys, you have to vote for the first post (thumb up) so this issue may be taken seriously. With only 2 votes everybody won't care even if it's an annoying issue.

**Comment by ElCondor1969** [comment]
A question to developer guys.
Will this issue be taken into account shortly and put into a milestone?
It's really a big flaw to reopen a multitude of files for work with the right encoding.
It's a shameful issue.

**Comment by hebu** [comment]
Yes, please fix the issues with "auto guess encoding". I tried to use this setting, but it tends to guess japanese encodings for utf-8 encoded markdown files or Windows-1252 encoded PowerShell skripts. And if you don't see it while saving, you'll lose all your special characters in the file (thanks to version control at this moment, but still you lose working time...)

**Comment by dudamoos** [comment]
This feature is actually really important for VHDL. The standard CHARACTER type in VHDL is defined by the ISO 8859-1 character set. Because CHARACTER is an enumeration and "each of the 191 graphic characters of this character set is denoted by the corresponding character literal" (VHDL LRM section 3.1.1.1), it is very difficult to use these values unless your editor uses the ISO 8859-1 character set for VHDL files. If you open up the std.STANDARD package (included in most EDA tools that support VHDL), you will see all of the values for CHARACTER replaced with unknown character symbols. Additionally, it will mangle files using these symbols when you save by converting all the non-ASCII characters to a UTF-8 bad character symbol, thus destroying data.

TL;DR: In order to use Visual Studio Code to edit VHDL files while still being able to use it to edit files in other languages encoded as UTF-8 (e.g. XML and HTML), this feature is extremely important.

**Comment by workflo** [comment]
Any news on this one? It's a real bummer :(

**Comment by FrancisFaure** [comment]
No news from Microsoft on my side ...

**Comment by freMea** [comment]
If only an extension could handle encoding according to a rule in **settings.json** or/and a comment at top of the file specifying the wanted priority encoding for this specific file.

But even then, VSCode doesn't even support some basic encoding such as **OEM** alias **cp 850** needed to display foreign chars and ASCII cosmetics in batch files. #23646 (vote for this issue too please). 

And yes the auto guessing encoding feature sucks comparing to other good old basic text editors.

😓

@FrancisFaure So, you said your team has offered a Pull Request to VSCode for this very issue and they didn't answer or valid?

**Comment by FrancisFaure** [comment]
Hi @freMea,

I just read your issue :
https://github.com/Microsoft/vscode/issues/23646
opened the march 30 of 2017

My issue :
https://github.com/Microsoft/vscode/issues/4891
opened the 3 April of 2016 

I think we have the same probleme (or near) = a folder can have files with different extension filenames : .html / .css / .asp / .aspx /.prg  / .bat ... 
but, if I well understand: vscode can not manage encoding per extension of filenames, but only by folders

for my language extension I proposed to add in "package.json" the encoding code or have an other similar solution.

about your sentence "So, you said your team has offered a Pull Request to VSCode for this very issue and they didn't answer or valid?" : 
My english is very poor and Google Trad: can not help me about your "Pull Request"  : I don't understood.

with best regards
Francis

**Comment by workflo** [comment]
There is hope: https://github.com/aadsm/jschardet/releases/tag/v1.5.0 :)

**Comment by Raydir** [comment]
voted for this issue.

**Comment by roffelsaurus** [comment]
"files.autoGuessencoding": true worked for me auto detecting win1252 on opening. The changes in @workflo link seem to at least have given a workaround for this issue.

**Comment by BlackOnion** [comment]
Voted for this issue.  
I like VSCode very much, but to not be able to set the encoding per file extension is a absolute no go. Even every normal text editor need this feature. I used it in the last millennium.

**Comment by bpasero** [comment]
Trying to understand the requirements here. So far I have seen various requests to associate a file encoding
* by language (https://github.com/Microsoft/vscode/issues/19890)
* by extension (this bug)
* by path (https://github.com/Microsoft/vscode/issues/23570)

It seems to me that being able to define the encoding by path is most powerful because it would allow to define the encoding for particular files in specific folders. And it would also allow to define for all files in all folders of a specific extension.

The beauty however of allowing to define the encoding by language is that we already have support for defining settings per language. The syntax is:

```json
"[markdown]":  {
  "editor.wordWrap": "on",
  "editor.quickSuggestions": false
}
```

And so I could envision a setting such as:

```json
"[markdown]":  {
  "files.encoding": "some encoding"
}
```

There is another related setting where you can define which extension maps to which language (`files.association`).

**Comment by Somnium7** [comment]
@bpasero I think that defining by path should allow wildcards to be able to match files by extension only. However, it could be harder to implement than detecting by language, because it needs new configuration syntax. For my personal needs, detection by language suffices. Let's see, what other people think.

**Comment by ElCondor1969** [comment]
I agree with @Somnium7 . Detection by language is the simplest and, at present, the fastest thing to implement.

**Comment by workflo** [comment]
I would like to add a fourth alternative to @bpasero's proposal:
It would be great if I could annotate an arbitrary text file and explicitly tell VSC which encoding to use. This was one of (X)Emacs' features I really liked:
`// -*- coding: latin1 -*-`

Emacs also allows overriding other settings like indentation mode, tab width and the like this way.

**Comment by bpasero** [comment]
I have pushed a change that allows to configure the encoding per language in settings (see https://github.com/Microsoft/vscode/issues/19890#issuecomment-328480143). Will be available in tomorrows insider release. I think we should see how it goes with this support before we introduce more powerful settings.

**Comment by BlackOnion** [comment]
Thank you, Benjamin.
I used the insider for a very small test and it worked for me.

**Comment by bpasero** [comment]
@FrancisFaure ok to close?

**Comment by FrancisFaure** [comment]
Hi @bpasero,
I don't had/took time to test the solution of this issue (I work on vscode on my free time and my english is poor) I think I can work on that: this week-end.
(If in the same directory we can have some files with differents encoding : ie .PRG in cp1252, .XML in utf8, ... and if I can change encoding for .PRG : that could be great)
with best regards
Francis

**Comment by bpasero** [comment]
@FrancisFaure yes, the current solution allows to change the encoding for language types. This is possible globally or per workspace folder.

**Comment by FrancisFaure** [comment]
Hi @bpasero
when you write "language types" = that mean by extension of filenames (ie : .PRG, .BAT, .CMD, .XML...) ?
is it a setting in  package.json of language extension ? other ? (if other have you a url link for read and test please) 
Regards

**Comment by bpasero** [comment]
@FrancisFaure this is an indirection: multiple extensions can be associated with one language. I suggest you give it a try and report back how it works for you.

**Comment by FrancisFaure** [comment]
Hi @bpasero
ok I will work on the subject this weekend,
Thank you
I'll let you informed after finding the links on VSCODE to define a "language type" to link the extension .PRG + tests
If you want, you can close the subject and if necessary reopen it if it does not work?
(Sorry for my bad English)
regards

**Comment by jhonrocha** [comment]
Hello @bpasero 
I just tried the privileged version. They're working perfectly on both global and workspace settings! 
Thank you, best regards

```
// ---- My settings.json ----
{
    "[asp]":{
      "files.encoding": "iso88591"
    },
    "[javascript]":{
      "files.encoding": "utf8"
    },
    "[html]":{
      "files.encoding": "utf8"
    }
}
```

**Comment by FrancisFaure** [comment]
Hi @bpasero,
I tested with VS Code Insider :

a) I added in my settings (stored in C:\Users\USER\AppData\Roaming\Code - Insiders\User\settings.json):
```
    "[vfp]":{
        "files.encoding": "windows1252"
    }
```
that work well.

b) I removed my settings of §a
and added in my package.json language :

```
	"contributes": {
		"configurationDefaults": {
		  "[vfp]": {
			"files.encoding": "windows1252"
		  }
		}...
```
and that work ! great ! it's what I needed :)
with best regards	
Francis

**Comment by Raydir** [comment]
works like a bomb

**Comment by Somnium7** [comment]
@bpasero 
Doesn't work search in multiple files without Ripgrep (haven't tested with it, don't need it).
I have files which need "maccenteuro" encoding, let assume it is Javascript.
With following configuration search works in all files except ".js".
`"search.useRipgrep": false`
With following configuration works in none files except ".js".
`"search.useRipgrep": false,`
`"files.encoding": "maccenteuro"`
With following configuration search should work for all files (seems logic), however, it works in all files except ".js" (like first case):
`"search.useRipgrep": false,`
`"[javascript]": {`
`"files.encoding": "maccenteuro"`
`}`
Note that when opening actual files encoding is working properly, only search in multiple files doesn't work.

Tested in VSCode version 1.17.0 on Windows 10

**Comment by buzzzzer** [comment]
@bpasero
Search in multiple files doesn't work with setting by language
Win10 1.17.0 release

**Comment by bpasero** [comment]
@roblourens is aware of [that](https://github.com/Microsoft/vscode/issues/4891#issuecomment-335370709) limitation afaik.

**Comment by Somnium7** [comment]
@roblourens Are you (or another person) going to fix this? Maybe a new issue should be created?

**Comment by roblourens** [comment]
This is the library that would have to support those encodings: https://github.com/hsivonen/encoding_rs

It's probably unlikely, as it's not a standard encoding.
